### DISCUSSION TITLE
HELP!- Unable to Play audio files from sd card from in build example - "player-sd-audiokit"

### DISCUSSION THREAD

**Discussion by Tanmay-J10** [pschatzmann/arduino-audio-tools#1631] [Q&A] [answered]
HI ,
First of big fan of yours, thanks for putting in you effort :)

i have been stuck on playing sd card example code, i am using esp32 dev module , along with uda1334a and sd card module and the following example code
[https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino](url)

i have tried both the uda1334a and sd card module individully and both work fine , but when i use the example it crashes with the follow logs-

[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 36 / logic: 0
[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 23 / logic: 0
[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 19 / logic: 0
[I] AudioPlayer.h : 338 - setVolume(0.700000)
[I] VolumeStream.h : 199 - setVolume: 0.700000 at 0
[I] VolumeStream.h : 199 - setVolume: 0.700000 at 1
[I] AudioTypes.h : 128 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed

and here is  the code  that i changed, (i have only changed the music directory, rest is the same)

#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"
#include "AudioLibs/AudioSourceSD.h" // or AudioSourceIdxSD.h
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
AudioBoardStream kit(AudioKitEs8388V1);
MP3DecoderHelix decoder;  // or change to MP3DecoderMAD
AudioPlayer player(source, kit, decoder);

void next(bool, int, void*) {
   player.next();
}

void previous(bool, int, void*) {
   player.previous();
}

void startStop(bool, int, void*) {
   player.setActive(!player.isActive());
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  // sd_active is setting up SPI with the right SD pins by calling 
  SPI.begin(23, 19, 18, 5);
  cfg.sd_active = false;
  kit.begin(cfg);

  // setup additional buttons 
  kit.addDefaultActions();
  kit.addAction(kit.getKey(1), startStop);
  kit.addAction(kit.getKey(4), next);
  kit.addAction(kit.getKey(3), previous);

  // setup player
  player.setVolume(0.7);
  player.begin();

  // select file with setPath() or setIndex()
  player.setPath("/music/1.mp3");
  //player.setIndex(1); // 2nd file
  `
}

void loop() {
  player.copy();
  kit.processActions();

}

**Comment by pschatzmann** [comment, marked as answer]
Don't use the [AudioBoardStream](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Boards) with an esp32 dev module !

If you want to use the AudioActions, read the corresponding Wiki about interactive Audio!

In the AudioSourceSD indicate the CS pin that you have connected with your SD!

**Comment by Tanmay-J10** [comment]
i am sry to disturb, i am a rookie , where exactly am i supposed to change the pin no 
![image](https://github.com/user-attachments/assets/4743ca06-9be6-49b5-9b3b-957c5e657b7f)

**Comment by pschatzmann** [comment]
Google the topic SPI, but on the ESP32 you can basically just use any pin: but you need to configure them, like you have done above.
In AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS); you specified the wrong CS Pin: It should be 5!
 
PIN_AUDIO_KIT_SD_CARD_CS is 13!

**Reply by Tanmay-J10** [reply]
![image](https://github.com/user-attachments/assets/1359fef9-b57b-4477-8965-1e2f23424c70)
i have tried it with cs = 5, but still it doesnt work , but while sd card test program, it works , heres the link to the sd card test [https://randomnerdtutorials.com/esp32-microsd-card-arduino/](url)

![image](https://github.com/user-attachments/assets/577fca42-37ff-4d4f-aa6f-1161ab3dd94e)here is the image of the test

**Comment by Tanmay-J10** [comment]
same problem again what do i do ?? 
😭

**Reply by pschatzmann** [reply]
I guess you still messed something up with the pin numbers. What are you doing differently to the test sketch ?
Check the pin numbers and definitions!

**Reply by Tanmay-J10** [reply]
i didnt change anything, i used you test code once, i got pin 15 error and then i changed it to pin 5 as you said, but still same thing happenning, thanks for you help though

**Reply by Tanmay-J10** [reply]
i used you exact example code also without modifications
, it doesnt work

**Reply by pschatzmann** [reply]
Did you delete your SPI.begin() with the messed up pins ?

**Reply by Tanmay-J10** [reply]
yes, but now it is constantly bootlooping 
![image](https://github.com/user-attachments/assets/35402556-3da5-4b1c-b4ec-cb90986bab17)

**Reply by Tanmay-J10** [reply]
i am getting this now ,



[I] I2SConfigESP32V1.h : 79 - signal_type: Digital
[I] I2SConfigESP32V1.h : 81 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32V1.h : 84 - use_apll: true
[I] I2SConfigESP32V1.h : 92 - pin_bck: 14
[I] I2SConfigESP32V1.h : 94 - pin_ws: 15
[I] I2SConfigESP32V1.h : 96 - pin_data: 22
[I] I2SESP32V1.h : 211 - tx: 22, rx: -1
[I] StreamCopy.h : 70 - buffer_size=1024
[I] AudioEncoded.h : 75 - virtual void audio_tools::EncodedAudioOutput::addNotifyAudioChange(audio_tools::AudioInfoSupport&)
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 158 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops

**Reply by pschatzmann** [reply]
So, it's working now...

**Reply by Tanmay-J10** [reply]
but i am unable to hear anything sir ,

**Reply by Tanmay-J10** [reply]
well anyways thank you for yourhelp , i dont want to waste any more of you time, but it aint working , Thanks A Lot!

**Reply by pschatzmann** [reply]
So you must have messed up with the i2s pins to your DAC

ps. for [your reference here you can find the information ](https://github.com/espressif/arduino-esp32/blob/master/libraries/SPI/src/SPI.h)in what sequence you need to pass the pins in SPI.begin(), should you decide to use some other assignments in the future.
